### PR TITLE
Admission_Syntax integration into public x509v3 header

### PR DESCRIPTION
**Pull request by matthiasballreich**
Admission_Syntax Integration into public x509v3 Header.
Third try because i am not so firm with git:/
Now it should work ;)

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by matthiasballreich** [comment]
What's with the pull request?

**Comment by matthiasballreich** [comment]
Fixed conflicts. Please merge.

**Comment by matthiasballreich** [comment]
@richsalz my mail is set to Matthias.Ballreich@outlook.de or i am wrong?
In the CLA i sent this mail is given, too.

**Comment by richsalz** [comment]
No, it's strange.  I'll clear the flag.  @levitte any clues?

**Comment by matthiasballreich** [comment]
very strange. When i run git config user.email i got as result "Matthias.Ballreich@outlook.de"
In the CLA i submit two mail addresses: "Matthias.Ballreich@outlook.de" and "Matthias.Ballreich@outlook.com".
Can you tell me which email you see?

**Comment by richsalz** [comment]
No, your email is fine (the outlook.de address); maybe our cla-bot has a bug.

**Comment by matthiasballreich** [comment]
ready for merge?

**Comment by richsalz** [comment]
Needs another @openssl reviewer.

**Comment by snhenson** [comment]
A question: this is placing a structure definition in a public header file. Do we really want to do that or have it opaque and use appropriate accessors? Yes there are *many* public structures in x509v3.h anyway but those were about before we wanted to make as much as possible opaque.

**Comment by richsalz** [comment]
I thought it was defined by a public standard and therefore it was okay.  That thought could be doubly-wrong, I admit :)

**Comment by FdaSilvaYY** [comment]
@snhenson : as x509 extensions are defined and describe by various RFC , and the related structure definition don't carry any internal or cache data , i don't see what is the value added  by making them  opaque ...

**Comment by t-j-h** [comment]
The value in making things opaque is that we can change  them in future in any way we want - and there are lots of reasons to want to have different implementation logic behind X509 structures.

**Comment by richsalz** [comment]
this is just i2d/d2i of a publically-defined structure.

**Comment by matthiasballreich** [comment]
so it's ready for merge?

**Comment by t-j-h** [comment]
@richsalz your argument is somewhat bogus - you could equally apply the argument to i2d_X509 - X509 is just a publicly defined structure. We should not dig the hole deeper - there is no benefit to forcing a single fixed implementation with no possibility of changes over time.

**Comment by richsalz** [comment]
I believe all we're going to do is convert to/from.  You disagree.  I'm willing to accede the point.  Which means that this PR is very incomplete, and now needs all the accessors at a minimum.  And perhaps a constructor.  So, @matthiasballreich it's not ready :(

**Comment by matthiasballreich** [comment]
okay but i don't unterstand the problem because the original PR with witch these structures are merged into openssl were full approved. In my PR i only moved some of these code parts into the public X509 header so that it is better accessible for the users.

I don't know how to convert these things in a non opaque structure, etc. So i can't do this.

**Comment by richsalz** [comment]
Ah yes, you're right.  @t-j-h it seems a chance to address your concern was missed.  Now what?

**Comment by t-j-h** [comment]
The huge difference here is that this is exposing these as part of the external interface in libcrypto - previously they were not "exported".

**Comment by richsalz** [comment]
well, it's no worse then the dozens of other i2d/d2i functions we have is it?

**Comment by wolftobias** [comment]
@snhenson I understand what you are saying, but I`m waiting for that feature since 19. September 2016. I`m waiting for this more than a year now. As fas as I know is currently no extension api made opaque yet. As @matthiasballreich already said we don`t know how you want to get the extensions opaque. Better would be if you as core developers make the api opaque to give us a chance to better know, see and understand what you are want us to do.
For me it would be nicer to use the admission feature now and change it later to a opaque api.

**Comment by snhenson** [comment]
We decided some time ago that new structures should be made opaque where possible unless there was a compelling reason not to do so. The reason why there aren't any extensions currently that use opaque structures is that no new ones have been added since that decision. Several ASN.1 structures (that aren't extensions) have been made opaque, for example X509, X509_REQ, X509_CRL.

Moving a structure to a public header means we can't change it in any 1.1.x version without breaking compatibility: the earliest it could be made opaque is 1.2.0.

Adding an opaque API isn't that difficult though. All you really need is accessors: get0/set0 for each field I think would be sufficient for this case. If that isn't clear enough let me know and I'll give you some examples of the kind of thing I mean.

**Comment by kaduk** [comment]
Reading through the history, it is hard to be certain, but it looks like the consensus is to leave this structure opaque and switch to accessors?  If so, presumably the "pending 2nd review" label should be removed.

**Comment by wolftobias** [comment]
@matthiasballreich How can we proceed. I understood the code you tried to get commited is in your branch, you cannot get it commited without the accessor the other guys asked for. When I start to make an implementation for accessors in my own branch, we loose your changes. 
@All How shall we implement the accessors? Shall we start a talk about this?

**Comment by wolftobias** [comment]
Today I call X509_get_ext_d2i to get an extension back, is this fine or shall we work with accessors here also?
```
	admission_syntax = (ADMISSION_SYNTAX*)X509_get_ext_d2i(x509, x509_add_admission_extensions(), critical, NULL);
```

**Comment by wolftobias** [comment]
Shall we implement these accessors like this below? @matthiasballreich We can do this together, I write the code and send this to you and you put this in your branch?

```
ADMISSIONS* X509v3_get_admissions(const ADMISSION_SYNTAX* x, int loc)
{
    if (x == NULL || OPENSSL_sk_num(x) <= loc || loc < 0)
        return NULL;
    else
        return (ADMISSIONS*)OPENSSL_sk_value((OPENSSL_STACK*)admission_syntax->contentsOfAdmissions, loc);
}
```

**Comment by matthiasballreich** [comment]
@wolftobias yes we can do this if i found the time for it. And yes you can send me the code which i have to include in my branch. But you or someone else has to tell me where i have to integrate the code. I don't know anything of the new opaque accessor stuff...

**Comment by wolftobias** [comment]
I think accessors are just functions which need to be exported like others too. They can be integrated here 
```
crypto/x509v3/v3_admis.c
```
I plan to make a accessor fuction for admission stack, one for profession stack coming from admission, and at least one for getting professionOIDs and professionItems, The same for setters and then it would fullfit my needs. 

@snhenson Is that assumption right?

**Comment by richsalz** [comment]
That is probably good enough if it meets your needs since you are the (first) one using it.
Don't forget the documentation, and the doc-nits target :)

**Comment by snhenson** [comment]
You don't need to add any wrapper to X509_get_ext_d2i(),

You shouldn't need any casts in the accessors at all. The names would be of the form STRUCTURE_{get0,set0}_FIELD. The get0 version is constant. So you have things like:
```
const GENERAL_NAME *ADMISSION_SYNTAX_get0_admissionAuthority(const ADMISSION_SYNTAX *as)
{
    return as->admissionAuthority;
}

void ADMISSION_SYNTAX_set0_admissionAuthority(ADMISSION_SYNTAX *as, GENERAL_NAME *aa)
{
    GENERAL_NAME_free(as->admissionAuthority);
    as->admissionAuthority = aa;
}
```

**Comment by wolftobias** [comment]
@matthiasballreich @snhenson I`m not the absolutly GIT expert, therefore I explain what I did. I need for the "accessors" feature the code from matthias, I`ve forked the openssl project from matthias github. Now I created a pull request for these changes. How can we proceed?

@matthiasballreich I`ve fixed the conflict for `util/libcrypto.num`

**Comment by wolftobias** [comment]
Can someone help me with this 
[https://travis-ci.org/matthiasballreich/openssl/jobs/278568545](Travis CI)

**Comment by wolftobias** [comment]
@richsalz I got no answer yet, just to confirm. After I provided my changes, is that branch now ready for commit or is something missing?

**Comment by matthiasballreich** [comment]
@wolftobias did you fix the travis build error? If not i maybe have some time the next days to look into it.

**Comment by wolftobias** [comment]
@matthiasballreich I`ve added the setter/getter functions in the code and in the `util/libcrypto.num` for export, then I saw there has been made an update, then I 've merged the `util/libcrypto.num` changes. After this the travis ran to the same point and failed you worked at above, but I couldn`t find the reason why. It seems it failed with some documentation checks, but I`m just guessing.

please check this job log
https://travis-ci.org/matthiasballreich/openssl/jobs/278568545

**Comment by wolftobias** [comment]
Means the last line in the travis `Done. Your build exited with 1.` OK or FAILURE? But the header is showing in red `Pull Request #1 added accessors`. In the log files I see no error. Can anybody help me further here?!

**Comment by matthiasballreich** [comment]
@wolftobias i migrated your code into mine

**Comment by wolftobias** [comment]
@richsalz Can you help us to find out what the problem with CI is?

**Comment by richsalz** [comment]
I am not sure it's your problem.  Looking at the full build log (searching for /// helps find failures quickly) I see
```
Test Summary Report
-------------------
../../_srcdist/test/recipes/70-test_sslrecords.t             (Wstat: 13 Tests: 3 Failed: 0)
  Non-zero wait status: 13
  Parse errors: Bad plan.  You planned 18 tests but ran 3.
```
Then looking at the log for that part of the test, I cannot see an error.  So I would ignore it.

**Comment by richsalz** [comment]
This code needs another review and then can be merged IMO

**Comment by wolftobias** [comment]
@snhenson Would you please be so kind to do the second review?

File: crypto/x509v3/v3_admis.c [hunk 1]
```diff
@@ -202,3 +201,113 @@ static int i2r_ADMISSION_SYNTAX(const struct v3_ext_method *method, void *in,
...
+}
+ 
+const STACK_OF(ADMISSIONS)* ADMISSION_SYNTAX_get0_contentsOfAdmissions(const ADMISSION_SYNTAX *as)
+{
+    return as->contentsOfAdmissions;
+}
+ 
+void ADMISSION_SYNTAX_set0_contentsOfAdmissions(ADMISSION_SYNTAX *as, STACK_OF(ADMISSIONS) *a)
+{
+    sk_ADMISSIONS_free(as->contentsOfAdmissions);
```
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: Need sk_ADMISSIONS_pop_free() here or there will be leaks.

File: crypto/x509v3/v3_admis.c [hunk 2]
```diff
@@ -202,3 +201,113 @@ static int i2r_ADMISSION_SYNTAX(const struct v3_ext_method *method, void *in,
...
+}
+ 
+const PROFESSION_INFOS* ADMISSIONS_get0_professionInfos(const ADMISSIONS *a)
+{
+    return a->professionInfos;
+}
+ 
+void ADMISSIONS_set0_professionInfos(ADMISSIONS *a, PROFESSION_INFOS *pi)
+{
+    sk_PROFESSION_INFO_free(a->professionInfos);
```
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: See above.

File: include/openssl/x509v3.h [hunk 3]
```diff
@@ -863,6 +863,68 @@ int X509v3_addr_validate_resource_set(STACK_OF(X509) *chain,
...
+void PROFESSION_INFO_set0_addProfessionInfo(PROFESSION_INFO *pi, ASN1_OCTET_STRING *aos);
+const NAMING_AUTHORITY* PROFESSION_INFO_get0_namingAuthority(const PROFESSION_INFO *pi);
+void PROFESSION_INFO_set0_namingAuthority(PROFESSION_INFO *pi, NAMING_AUTHORITY *na);
+const STACK_OF(ASN1_STRING)* PROFESSION_INFO_get0_professionItems(const PROFESSION_INFO *pi);
+void PROFESSION_INFO_set0_professionItems(PROFESSION_INFO *pi, STACK_OF(ASN1_STRING) *as);
+const STACK_OF(ASN1_OBJECT)* PROFESSION_INFO_get0_professionOIDs(const PROFESSION_INFO *pi);
+void PROFESSION_INFO_set0_professionOIDs(PROFESSION_INFO *pi, STACK_OF(ASN1_OBJECT) *po);
+const ASN1_PRINTABLESTRING* PROFESSION_INFO_get0_registrationNumber(const PROFESSION_INFO *pi);
+void PROFESSION_INFO_set0_registrationNumber(PROFESSION_INFO *pi, ASN1_PRINTABLESTRING *rn);
+
```
Comment (snhenson) [inline include/openssl/x509v3.h]: This part needs removing: we don't want the structure definition in a public header.

File: crypto/x509v3/v3_admis.c [hunk 4]
```diff
@@ -14,46 +14,45 @@
 #include <openssl/asn1t.h>
 
 #include <openssl/x509v3.h>
-
+
```
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: Not sure what is going on here and elsewhere? Seems some spaces are creeping in.

**Comment by snhenson** [comment]
The formatting is also all over the place. TBH it would be quicker if I just extracted the relevant parts of this PR and submitted them as a fresh PR.

File: crypto/x509v3/v3_admis.c (hunk 1 above)
Comment (wolftobias) [reply crypto/x509v3/v3_admis.c]: done

File: crypto/x509v3/v3_admis.c (hunk 2 above)
Comment (wolftobias) [reply crypto/x509v3/v3_admis.c]: done

File: crypto/x509v3/v3_admis.c (hunk 4 above)
Comment (wolftobias) [reply crypto/x509v3/v3_admis.c]: done

File: include/openssl/x509v3.h (hunk 3 above)
Comment (wolftobias) [reply include/openssl/x509v3.h]: But when I remove these functions, we'll not be able to set the oids and the registration number. But these are import data to set! We need to set these!

**Comment by snhenson** [comment]
See #4474 for an alternative version based on this.

**Comment by richsalz** [comment]
Please see #4724.

**Pull request closed without merging**